lib/Sema/CSApply.cpp

```diff
@@ -2406,8 +2411,90 @@ namespace {
 
     // SWIFT_ENABLE_TENSORFLOW
     Expr *visitTFOp(ObjectLiteralExpr *expr) {
+      auto &ctx = cs.TC.Context;
+
+      auto diagnose = [&](SourceLoc loc, const Twine &message) {
+        cs.TC.diagnose(loc, diag::invalid_tfop, message.str());
+      };
+
       // All #tfop operands are RValues.
       expr->setArg(cs.coerceToRValue(expr->getArg()));
+
+      // We do most of the typechecking for #tfop here in CSApply instead of
+      // generating constraints for #tfop in CSGen because #tfop allows many
+      // different types of arguments and results that are inconvenient or
+      // impossible to generate constraints for. It's much easier to check #tfop
+      // after we know what all the argument and result types are.
+
+      // Check arguments.
+      auto *args = dyn_cast<TupleExpr>(expr->getArg());
+      if (!args) {
+        // CSGen has already ensured that the argument is a string in the
+        // single argument case.
+        return expr;
+      }
+
+      // CSGen has already ensured that the first argument is a string, so skip
+      // the first argument.
+      for (unsigned i = 1, e = args->getNumElements(); i != e; ++i) {
+        auto argExpr = args->getElement(i);
+        auto argType = cs.getType(argExpr);
+        auto argLoc = argExpr->getLoc();
+        auto argNameAndLowering = tf::GraphOperationInfo::decodeArgumentName(
+            args->getElementName(i).str());
+        if (!argNameAndLowering) {
+          diagnose(argLoc, "argument name has invalid modifier");
+          return nullptr;
+        }
+        auto argLowering = argNameAndLowering->second;
+        switch (argLowering) {
+        case tf::GraphOperationInfo::ArgumentLowering::Input:
+          // CSGen has already ensured that all inputs conform to
+          // TensorArrayProtocol.
+          break;
+        case tf::GraphOperationInfo::ArgumentLowering::NormalAttribute:
+          if (atc.classifyNormalAttribute(argType) ==
+              tf::AttributeTypeClassifier::Normal::Unsupported) {
+            diagnose(argLoc, "attribute requires Bool, Int64, Double, Float, "
```

Comment:
[nit] is there a way and would it be useful to define the "attribute requires" part of the error string within `tf::AttributeTypeClassifier`, so that its content is jointly maintained with the content of the `Normal` enum class? same for the other enum classes below.

Reply:
I can make a static string constant in AttributeTypeClassifier.